### PR TITLE
RADIO_RC_CHANNELS message

### PR DESCRIPTION
**Pull request by olliw42**
Suggestion for a new RADIO_RC_CHANNELS message.

The backgorund is given in the issue https://github.com/mavlink/mavlink/issues/1921.

If generally approved, the details of it may be discused here in this PR.

### CODE REVIEW

**Comment by olliw42** [comment]
list of points which occured in discussions
- **message id**: The message is currently located in storm32.xml with an id in the storm32.xml's range. It doesn't smell correct to keep that id when moved to development. So I changed it. One could use any free id in the mavlink2 range as onenromally would do. In this PR I however suggest to use a gap in the mavlink1 area, which could host a number of releated messages. Personally I totally do not care, so fill happily follow any suggestion. It just needs to be decided.
- **target fields**: The message has no target fields, and it is proposed that the desired targeting is achieved by special code in the routers, similar to how the existing RADIO_STATUS is handled. This can be much debated, with pros and cons, but a decission needs to be made.
- **source system id**: The receivers sending out this message are supposed to have a source system id different from the vehicle/system they are connected to. Typically 51. This is similar to SiK radios and alike.
- The receivers are normally not supposed to send HEARTBEAT messages (unless they want to support microservices, like e.g. parameters)
 
The message originally was sugested to carry 24 channels max. I have extended to 32 channels, since due to zero-padding it doesn't cost anything but may make it super future proof.

**Comment by olliw42** [comment]
@hamishwillee maybe I can direct your attention towards this PR for a moment. Last week there had been some encouraging activity on the ArduPilot side with respect of implementing the proposed RADIO_RC_CHANNELS message, and the ArduPilot devs seem to have generally agreed to the concept: https://github.com/ArduPilot/ardupilot/pull/24577#pullrequestreview-1751596027, https://github.com/ArduPilot/ardupilot/pull/24577#pullrequestreview-1751592965. While activity has lowered a bit, I think it could be the appropriate time to discuss and settle the topic of which _**message id**_ to choose.

Since besides RADIO_RC_CHANNELS there would be few more messages in the "RADIO series", it could make sense to look for slots of 5 messages or so (I currently do see 4 messages).  

The PR currently suggests a message id in the MAVLink v1 range, namely 95, as there happens to be a free slot from 93 - 99.

The canonical approach would be to occupy the "next" slot in the MAVLink v2 area, which seems to be 420 currently.

All I would want you to decide is if 95 is ok to choose, with the understanding that it essentially would reserve the slot 95-99, or if I should change the PR to use 420?

Personally I have no real opinion here, It just needs to be decided. :)

THX. Olli

**Comment by hamishwillee** [comment]
IMO the range should change to something above 256. It doesn't have to be 420 - but if that looks like a good free block then sure.

**Comment by olliw42** [comment]
ok, I'll change. Many thx!!

File: message_definitions/v1.0/development.xml [hunk 1]
```diff
@@ -481,6 +490,16 @@
...
+    <message id="420" name="RADIO_RC_CHANNELS">
+      <description>Radio channels. Supports up to 32 channels. Channel values are in centered 13 bit format. Range is [-4096,4096], center is 0. Conversion to PWM is x * 5/32 + 1500. 
+	    Should be emitted only by components with component id MAV_COMP_ID_TELEMETRY_RADIO.
+		The message should be handled, e.g. by routers, as if it had target fields with target_system = system id of system the radio receiver is connected to and target_component = 0. 
+		</description>
+      <field type="uint8_t" name="count">Total number of RC channels being received. This can be larger than 24, indicating that more channels are available but not given in this message.</field>
+      <field type="uint16_t" name="flags" enum="RADIO_RC_CHANNELS_FLAGS" display="bitmask">Radio rc channels status flags.</field>
+      <extensions/>
+      <field type="int16_t[32]" name="channels">RC channels. Channels above count should be set to 0, to benefit from MAVLink's zero padding.</field>
+    </message>
```
Comment (peterbarker) [inline message_definitions/v1.0/development.xml:502]: No RSSI or LQ?  People really like having this!

**Review by peterbarker — CHANGES_REQUESTED** [review]

File: message_definitions/v1.0/development.xml (hunk 1 above)
Comment (olliw42) [reply message_definitions/v1.0/development.xml:502]: NO!!

I'm a bit inclined to direct you to the fuller discussion https://github.com/mavlink/mavlink/issues/1921, but you sure deserve a briefer response. 

It's IMHO a strength of the suggested concept to separate the rc data from the link stats data, for a number of reasons. The link stats data is carried in the proposed RADIO_LINK_STATS message, the upstream PR is https://github.com/mavlink/mavlink/pull/1920. 

The intention is clearly that once the RADIO_RC_CHANNELS thing should have been accepted, to also raise the PR to bring in the RADIO_LINK_STATS. The code is available since long. The proposed RADIO_LINK_STATS message has seen btw already substantial discussion, since it was the template for a similar DroneCAN message.
So, in short, the link stats PR will follow.

Comment (peterbarker) [reply message_definitions/v1.0/development.xml:502]: OK, had a quick read of the relevant section in the linked issue - that's a fair argument, thanks!

**Comment by olliw42** [comment]
@hamishwillee 
FYI, the topic of this PR has made progress: 
The RADIO_RC_CHANNELS message and potential implementation in ArduPilot has been dicussed intensively over at ArduPilot, and there is now a PR in the ardupilot mavlink repository which from "our" point of view might be short of being accepted.

So, it would be great to also get your input. 

I kindly want to ask you to look over it and comment in the ArduPilot repo: https://github.com/ArduPilot/mavlink/pull/343
It would safe me from substantial double-burden - once the message has been merged to the ArduPilot mavlink repo I will update this PR here to be in sync, so that the two repos don't stray away. Hope you find this ok.

The main changes to the message proposed here had been
- adding target fields
- adding a time stamp for the when the rc data was valid the last time
- more comments in the description

THX

File: message_definitions/v1.0/development.xml [hunk 2]
```diff
@@ -481,6 +490,16 @@
...
         The receiving end can infer timestamp format (since 1.1.1970 or since system boot) by checking for the magnitude of the number.</field>
     </message>
+    <message id="420" name="RADIO_RC_CHANNELS">
+      <description>Radio channels. Supports up to 32 channels. Channel values are in centered 13 bit format. Range is [-4096,4096], center is 0. Conversion to PWM is x * 5/32 + 1500. 
+	    Should be emitted only by components with component id MAV_COMP_ID_TELEMETRY_RADIO.
+		The message should be handled, e.g. by routers, as if it had target fields with target_system = system id of system the radio receiver is connected to and target_component = 0. 
+		</description>
+      <field type="uint8_t" name="count">Total number of RC channels being received. This can be larger than 24, indicating that more channels are available but not given in this message.</field>
+      <field type="uint16_t" name="flags" enum="RADIO_RC_CHANNELS_FLAGS" display="bitmask">Radio rc channels status flags.</field>
+      <extensions/>
```
Comment (hamishwillee) [inline message_definitions/v1.0/development.xml:500]: Given the id is changing anyway, the message is breaking. So we don't need extensions.

Comment (hamishwillee) [reply message_definitions/v1.0/development.xml:500]:
```suggestion
```

File: message_definitions/v1.0/development.xml [hunk 3]
```diff
@@ -481,6 +490,16 @@
       <field type="uint64_t" name="time_usec" units="us">Timestamp (UNIX Epoch time or time since system boot).
         The receiving end can infer timestamp format (since 1.1.1970 or since system boot) by checking for the magnitude of the number.</field>
     </message>
+    <message id="420" name="RADIO_RC_CHANNELS">
+      <description>Radio channels. Supports up to 32 channels. Channel values are in centered 13 bit format. Range is [-4096,4096], center is 0. Conversion to PWM is x * 5/32 + 1500. 
+	    Should be emitted only by components with component id MAV_COMP_ID_TELEMETRY_RADIO.
+		The message should be handled, e.g. by routers, as if it had target fields with target_system = system id of system the radio receiver is connected to and target_component = 0. 
+		</description>
+      <field type="uint8_t" name="count">Total number of RC channels being received. This can be larger than 24, indicating that more channels are available but not given in this message.</field>
```
Comment (hamishwillee) [inline message_definitions/v1.0/development.xml:498]: ?
```suggestion
      <field type="uint8_t" name="count">Total number of RC channels being received. This can be larger than 32, indicating that more channels are available but not given in this message.</field>
```

Comment (hamishwillee) [reply message_definitions/v1.0/development.xml:498]: Argg. I'll look on the one on ArduPilot.

**Comment by hamishwillee** [comment]
@olliw42 Re https://github.com/mavlink/mavlink/pull/1919#issuecomment-1914268242 I have put in some comments, which hopefully make sense. The little picture is
- Extensibility - how many channels
- How does the timestamp and RC-lost behaviour work

The big picture is "can this be a single path for analog manual control input" with Joystick. Essentially addressing James comments in https://github.com/mavlink/mavlink/issues/1921#issuecomment-1331303555

Other than that, we probably need review from someone like @ThomasDebrunner and @julianoes - but let's see if my questions make any sense to you first :-)

**Comment by olliw42** [comment]
MANY thx for looking at it
I'll comment in the ArduPilot PR https://github.com/ArduPilot/mavlink/pull/343#issuecomment-1918417134, if you allow.

**Comment by hamishwillee** [comment]
Of course. Looks like broader interest :-)

**Comment by olliw42** [comment]
@hamishwillee brings in the changes to RADIO_RC_CHANNELS discussed in https://github.com/ArduPilot/mavlink/pull/343

also removes all other related enums/messages from the storm32.xml, as it seems it's all going to come :)

**Comment by olliw42** [comment]
ups, error ... it seems you had added minValue/maxValue fields to the message which are in ardupilot's repo but not yet in the schema here ... what to do? I guess you need to update the schema ...

**Comment by olliw42** [comment]
@hamishwillee removed the offending minValue,maxValue attributes, so that this message can be merged as soon as possible, I think the minValue,maxValue can be reinserted once the schema is done.

**Comment by hamishwillee** [comment]
The schema should have come back in today. If you can add back I'll merge tomorrow.

**Comment by olliw42** [comment]
done :)

**Comment by olliw42** [comment]
THX!
